src/audio/midi.rs

```diff
@@ -0,0 +1,146 @@
+use crate::Result;
+use anyhow::anyhow;
+use midir::{MidiInput, MidiInputPort};
+use num_traits::Num;
+use std::sync::mpsc::{self, Receiver};
+use std::thread;
+use std::time::Duration;
+use std::{convert::TryFrom, sync::mpsc::Sender};
+use wmidi::{Channel, ControlFunction, ControlValue, MidiMessage};
+
+/// Listens on the provided port, and sends MIDI messages over a channel. Returns the receiver of that channel.
+pub fn listen_for_input(port_name: &str) -> Result<Receiver<MidiMessage<'static>>> {
+    let port = port(port_name)?;
+
+    println!("MIDI input: {}", port_name);
+
+    let (sender, receiver) = mpsc::channel();
+    handle_messages(port, sender);
+
+    Ok(receiver)
+}
+
+pub fn port_names() -> Result<Vec<String>> {
+    let midi_input = midi_input()?;
+    midi_input
+        .ports()
+        .iter()
+        .map(|port| midi_input.port_name(port).map_err(|e| e.into()))
+        .collect()
+}
+
+/// Get the last value for a `ControlFunction` from a list of messages.
+pub fn latest_control_value(
+    messages: &[MidiMessage<'static>],
+    channel: Channel,
+    control_function: ControlFunction,
+) -> Option<ControlValue> {
+    messages.iter().rev().find_map(|message| match message {
+        MidiMessage::ControlChange(ch, function, value)
+            if ch == &channel && function == &control_function =>
+        {
+            Some(*value)
+        }
+
+        _ => None,
+    })
+}
+
+/// Maps a control value (0-127) into a range.
+/// eg. `control_value_in_range(50, 100, 64) == 75`
+pub fn control_value_in_range<T: Num + From<u8> + Copy>(min: T, max: T, value: ControlValue) -> T {
+    let value: T = from_control_value(value);
+    value * (max - min) / from_control_value(ControlValue::MAX) + min
+}
+
+fn midi_input() -> Result<MidiInput> {
+    Ok(MidiInput::new("Input")?)
+}
+
+fn port(name: &str) -> Result<MidiInputPort> {
+    let names = port_names()?;
+    let midi_input = midi_input()?;
+
+    midi_input
+        .ports()
+        .into_iter()
+        .find(|port| midi_input.port_name(&port) == Ok(name.into()))
+        .ok_or_else(|| {
+            anyhow!(
+                "Could not find a MIDI port with name '{}'. Available ports are:\n{}",
+                name,
+                names.join("\n")
+            )
+        })
+}
+
+fn handle_messages(port: MidiInputPort, sender: Sender<MidiMessage<'static>>) {
```

Comment:
When you see things that look like type parameters that start with `'` (eg. `'static`), those are lifetimes. Rust is silently adding lifetime annotations everywhere when it can figure them out, but in some cases you need to be specific. I'm sure it's explained much better in the book, but the basic idea is that if you have a function like this:

```rust
fn do_something(x: SomeStruct) -> &str {
  &x.some_field
}
```

it returns a reference to something that's owned by the `x` struct, so that reference is only going to be valid for as long as `x` itself is valid. In this case, Rust can determine the lifetime of the reference for you, but if you annotated it yourself, it'd look like this:

```rust
fn do_something<'a>(x: SomeStruct<'a>) -> &<'a>str {
  &x.some_field
}
```

(The convention is to use things like `'a` for the lifetime names 🤷🏻 )

`MidiMessage` by default internally uses a reference to the bytes it parsed, so it's only usable as long as those bytes are still around. I don't care about the bytes, so I needed to convert them to another lifetime, which I do with `drop_unowned_sysex` below. I'm kind of fuzzy on what `'static` actually means (something kind of non-obvious, I think).

I wouldn't worry too much about this stuff right now. It doesn't come up a ton.